src/compute/src/logging/timely.rs

```diff
@@ -526,3 +327,403 @@ fn create_address_row(id: usize, worker: WorkerIdentifier, address: &[usize]) ->
+    operates: OutputBuffer<'a, 'b, OperateDatum>,
+    channels: OutputBuffer<'a, 'b, ChannelDatum>,
+    addresses: OutputBuffer<'a, 'b, AddressDatum>,
+    parks: OutputBuffer<'a, 'b, ParkDatum>,
+    messages_sent: OutputBuffer<'a, 'b, MessageDatum>,
+    messages_received: OutputBuffer<'a, 'b, MessageDatum>,
+    schedules_duration: OutputBuffer<'a, 'b, usize>,
+    schedules_histogram: OutputBuffer<'a, 'b, ScheduleHistogramDatum>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct OperateDatum {
+    id: usize,
+    name: String,
+}
+
+impl Columnation for OperateDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct ChannelDatum {
+    id: usize,
+    source: (usize, usize),
+    target: (usize, usize),
+}
+
+impl Columnation for ChannelDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct AddressDatum {
+    id: usize,
+    address: Vec<usize>,
+}
+
+impl Columnation for AddressDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct ParkDatum {
+    duration_pow: u128,
+    requested_pow: Option<u128>,
+}
+
+impl Columnation for ParkDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct MessageDatum {
+    channel: usize,
+    worker: usize,
+}
+
+impl Columnation for MessageDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct ScheduleHistogramDatum {
+    operator: usize,
+    duration_pow: u128,
+}
+
+impl Columnation for ScheduleHistogramDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+/// Event handler of the demux operator.
+struct DemuxHandler<'a, 'b, 'c> {
+    /// State kept by the demux operator.
+    state: &'a mut DemuxState,
+    /// Demux output buffers.
+    output: &'a mut DemuxOutput<'b, 'c>,
+    /// The logging interval specifying the time granularity for the updates.
+    logging_interval_ms: u128,
+    /// The number of timely workers.
+    peers: usize,
+    /// The current event time.
+    time: Duration,
+    /// A capability usable for emitting outputs.
+    cap: &'a InputCapability<Timestamp>,
+}
+
+impl DemuxHandler<'_, '_, '_> {
+    /// Return the timestamp associated with the current event, based on the event time and the
+    /// logging interval.
+    fn ts(&self) -> Timestamp {
+        let time_ms = self.time.as_millis();
+        let interval = self.logging_interval_ms;
+        let rounded = (time_ms / interval + 1) * interval;
+        rounded.try_into().expect("must fit")
+    }
+
+    /// Handle the given timely event.
+    fn handle(&mut self, event: TimelyEvent) {
```

Comment:
We should be checking if the compiler inlines the `handle_*` functions. Maybe it'd be good to add an `#[inline(always)]` annotation to all functions, or explicit `#[inline(never)]` for those that are rarely called. I haven't thought too much which ones that might be!